src/services/game-logic/GameEngine.ts

```diff
@@ -12,22 +12,52 @@ class GameEngine {
 
     getLegalMoves = (square: Square): Square[] => {
         const piece = this.board.getPiece(square);
-        let pieceMoves = piece?.getPossibleMoves() ?? [];
-        return pieceMoves
-            .filter(this.isOnBoard)
-            .filter((square) => !this.isPawnAndBlocked(square, piece))
-            .filter((square) => !this.isOccupiedBySameColor(square, piece));
+        let legalMoves = piece?.getPossibleMoves().filter(this.isOnBoard) ?? [];
+        if (piece?.name === PieceNames.KING)
+            legalMoves = legalMoves.filter((square) => !this.isCastlingIllegal(square, piece));
+        if (piece?.name === PieceNames.PAWN) legalMoves = legalMoves.filter((square) => !this.board.getPiece(square));
+        return legalMoves.filter((square) => !this.isOccupiedBySameColor(square, piece));
     };
 
+    public movePiece(location: Square, destination: Square): void {
+        //can be useful when saving moves
+        const piece = this.board.getPiece(location);
+        this.board.movePiece(location, destination);
+        if (piece?.name === PieceNames.KING) this.performCastling(location, destination);
+    }
+
+    private performCastling(location: Square, destination: Square): void {
+        if (location.column - destination.column === -2) {
+            this.board.movePiece(new Square(location.row, 7), new Square(location.row, 5));
+        } else if (location.column - destination.column === 2) {
+            this.board.movePiece(new Square(location.row, 0), new Square(location.row, 3));
+        }
+    }
+
+    private isCastlingIllegal = (square: Square, piece: Piece): Boolean => {
+        let rook = null;
+        let rookTarget = null;
+        if (square.column - piece.position.column === 2) {
+            rook = this.board.getPiece(new Square(piece.position.row, 7));
```

Comment:
sugestia: można te ify / wnętrza ifów wyekstrachować do metod z ładnymi nazwami.
+ czytając kod nie wiem dlaczego tam jest 2, czy 7, czy 3, to są tzn magic numbers, można to też ponazywać typu whiteRookStartingColumn  czy coś w ten deseń